src/Features/LanguageServer/Protocol/Handler/CodeActions/CodeActionHelpers.cs

```diff
@@ -24,56 +23,128 @@ namespace Microsoft.CodeAnalysis.LanguageServer.Handler.CodeActions
+            // Filter out code actions with options since they'll show dialogs and we can't remote the UI and the options.
+            => suggestedAction.OriginalCodeAction is CodeActionWithOptions
+            // Skip code actions that requires non-document changes.  We can't apply them in LSP currently.
+            // https://github.com/dotnet/roslyn/issues/48698
+            || suggestedAction.OriginalCodeAction.Tags.Contains(CodeAction.RequiresNonDocumentChange);
+
+        /// <summary>
+        /// Generate the matching code actions for <paramref name="suggestedAction"/>. If it contains nested code actions, flatten them into an array.
+        /// </summary>
+        private static LSP.CodeAction[] GenerateCodeActions(
+            CodeActionParams request,
+            IUnifiedSuggestedAction suggestedAction,
+            LSP.CodeActionKind codeActionKind,
+            string currentTitle = "")
+        {
+            if (!string.IsNullOrEmpty(currentTitle))
+            {
+                // Adding a delimiter for nested code actions, e.g. 'Suppress or Configure issues|Suppress IDEXXXX|in Source'
+                currentTitle += '|';
+            }
+
+            var codeAction = suggestedAction.OriginalCodeAction;
+            currentTitle += codeAction.Title;
 
-                    codeActions.Add(GenerateVSCodeAction(
-                        request, documentText,
-                        suggestedAction: suggestedAction,
-                        codeActionKind: GetCodeActionKindFromSuggestedActionCategoryName(set.CategoryName!),
-                        setPriority: set.Priority,
-                        applicableRange: set.ApplicableToSpan.HasValue ? ProtocolConversions.TextSpanToRange(set.ApplicableToSpan.Value, documentText) : null,
-                        currentSetNumber: currentSetNumber,
-                        currentHighestSetNumber: ref currentHighestSetNumber));
+            var diagnosticsForFix = GetApplicableDiagnostics(request.Context, suggestedAction);
+
+            using var _ = ArrayBuilder<LSP.CodeAction>.GetInstance(out var builder);
+            if (suggestedAction is UnifiedSuggestedActionWithNestedActions unifiedSuggestedActions)
+            {
+                foreach (var actionSet in unifiedSuggestedActions.NestedActionSets)
+                {
+                    foreach (var action in actionSet.Actions)
+                    {
+                        builder.AddRange(GenerateCodeActions(
+                            request,
+                            action,
+                            codeActionKind,
+                            currentTitle));
+                    }
                 }
             }
+            else
+            {
+                builder.Add(new LSP.CodeAction
+                {
+                    // Add extra space because this is shown to user.
+                    Title = currentTitle.Replace("|", " | "),
```

Comment:
Is there any reason why we don't want spaces in `currentTitle`?

Reply:
Because `currrentTitle` is used as a key when resolve the code action.
See here https://github.com/dotnet/roslyn/blob/0c70822041b9b0d12f9e5eac97a2c45948a4612c/src/Features/LanguageServer/Protocol/Handler/CodeActions/CodeActionHelpers.cs#L297